### DISCUSSION TITLE
Static site, API calls with dynamic URLs

### DISCUSSION THREAD

**Discussion by imduchy** [nuxt/nuxt#9825] [Questions (Nuxt 2)]
Version: `nuxt: "2.15.7"`

Hello. 👋🏻 I don't find documentation very informative when it comes to the topic of making API calls from a static website. Yes, it's clearly written that API calls will be called during generation and results will be cached in a payload file, but what about API calls that require dynamic query parameters?

### My scenario
I have a simple static application (`target: 'static'`) hosted in Azure's blob storage. My application consists of 4 pages without dynamic routes. On my index page, in one of components, I'm using the `fetch` hook to get the latest data from my API.

```typescript
async fetch() {
    this.competition = await this.$axios.$get('/groups/competition', {
      params: {
        group: this.$store.state.group._id,
        team: this.followedTeam.apiId,
        season: this.upcomingGame.season,
        competition: this.upcomingGame.competitionId,
      },
    });
  },
```

As you can see, the API call requires some dynamic values (mostly coming from vuex store). I don't see how this API call could be made during site generation as it requires a logged-in user and data populated in the store.

### Question
I see that every time I navigate to the index page, this `fetch` hook is executed and an API call is made. Why is that happening? Is it because Nuxt can't fetch the data during generation? Is it because of something else? And if I had an endpoint that doesn't require dynamic query parameters, how can I make sure that an API call is made on every page load and a user gets always up-to-date data (instead of re-generating the whole site)?

**Comment by imduchy** [comment]
bump

**Reply by r9mp** [reply]
@Duchynko Hi there.
First of all: do you have a `server: true` or a `server: false` in your nuxt.config.js?

**Reply by imduchy** [reply]
Hello @r9mp 

I don't have `server` defined in my `nuxt.config.js`. Also, the `server` property doesn't accept boolean (according to the [documentation](https://nuxtjs.org/docs/configuration-glossary/configuration-server)). 

Did you mean the `ssr` property? I don't have that set either.

**Reply by r9mp** [reply]
@Duchynko  Oups, sorry, I was speaking about the ssr property: https://nuxtjs.org/docs/configuration-glossary/configuration-ssr#the-ssr-property

**Reply by r9mp** [reply]
@Duchynko 

I made few tests as what your saying seemed akward.
I created a project with:
```
ssr: true,
target: 'static',
```
Then `yarn generate` and `nuxt serve`
=> fetch is executed on server side, never client side.

I also tried with errors. But same result.

So, I'll need a repository to make further investigations.
=> Just one question: if you do a `console.log` in your fetch function, does it appears on server (on `nuxt generate`) and client?

Anyway. To answer your question, if this.$store.state.group._id is not yet defined, it will fail on ssr and data won't be available (or be incomplete, depending on your WS). You can use `fetchOnServer: false` to force fetch to be executed on client only.
If you don't want to regenerate you app at each data update (this can be automatized, but there is always a delay between data update and deployment), or fetch on client side, you have to use SSR without a target set to static.

**Reply by imduchy** [reply]
> If you don't want to regenerate you app at each data update (this can be automatized, but there is always a delay between data update and deployment), or fetch on client side, you have to use SSR without a target set to static.

That's what I had been thinking too. But apparently, my statically hosted application can make API calls on the client-side using dynamic values stored in vuex.

> Just one question: if you do a console.log in your fetch function, does it appears on server (on nuxt generate) and client?

Here is the full `generate` output from the console. I had added a `console.log` statement into the `fetch` method, but it was not printed on the server-side. On the client-side, it was printed out to the console (in a browser).

```
> nuxt-ts generate

ℹ Doing webpack rebuild because assets/variables.scss modified                         20:18:54
ℹ Production build                                                                     20:18:55
ℹ Bundling for server and client side                                                  20:18:55
ℹ Target: static                                                                       20:18:55
ℹ Using components loader to optimize imports                                          20:18:55
ℹ Discovered Components: node_modules/.cache/nuxt/components/readme.md                 20:18:55
✔ Builder initialized                                                                  20:18:55
✔ Nuxt files generated                                                                 20:18:55
ℹ Starting type checking service...                                    nuxt:typescript 20:18:58

✔ Client
  Compiled successfully in 32.03s

✔ Server
  Compiled successfully in 8.87s


Hash: 489f03fc61bd5e3832a4
Version: webpack 4.46.0
Time: 32031ms
Built at: 11/23/2021 20:19:30
                         Asset      Size  Chunks                                Chunk Names
../server/client.manifest.json  21.1 KiB          [emitted]                     
                    0236b05.js  55.2 KiB       2  [emitted] [immutable]         vendors/pages/admin/pages/table
                    117faf5.js    11 KiB      12  [emitted] [immutable]         pages/password
                    1dcfd00.js    22 KiB       8  [emitted] [immutable]         components/user-bet
                    2de22ac.js  2.45 KiB      14  [emitted] [immutable]         runtime
                    54577bd.js  2.25 KiB      13  [emitted] [immutable]         pages/table
                    56165e8.js  14.6 KiB   10, 6  [emitted] [immutable]         pages/index
                    5a42758.js  11.3 KiB      11  [emitted] [immutable]         pages/login
                    6c9a495.js  22.3 KiB      16  [emitted] [immutable]         vendors/pages/admin
                    6fc8d4f.js  1.34 KiB       6  [emitted] [immutable]         components/current-bet
                    758a47e.js  38.6 KiB       1  [emitted] [immutable]         vendors/components/bet-input/pages/admin/pages/index/pages/table
                    8665f46.js   270 KiB       4  [emitted] [immutable]  [big]  commons/app
                    9087225.js  1.01 MiB      17  [emitted] [immutable]  [big]  vendors/pages/index
                      LICENSES  17.3 KiB          [emitted]                     
                    b65cd65.js  58.9 KiB       0  [emitted] [immutable]         vendors/components/bet-input/components/current-bet/pages/admin/pages/index/pages/login/pages/passwo/86580b28
                    be4e02d.js  11.6 KiB       9  [emitted] [immutable]         pages/admin
                    c9f9d74.js   598 KiB      15  [emitted] [immutable]  [big]  vendors/app
                    d2da53d.js  10.4 KiB       3  [emitted] [immutable]         app
                    d95b23e.js   6.3 KiB       5  [emitted] [immutable]         components/bet-input
                    dba5476.js  19.8 KiB       7  [emitted] [immutable]         components/upcoming-game
 + 2 hidden assets
Entrypoint app = 2de22ac.js 8665f46.js c9f9d74.js d2da53d.js

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  8665f46.js (270 KiB)
  c9f9d74.js (598 KiB)
  9087225.js (1.01 MiB)

Hash: 9d43e43e42fe2b15e1d8
Version: webpack 4.46.0
Time: 8871ms
Built at: 11/23/2021 20:19:38
                      Asset      Size         Chunks             Chunk Names
    components/bet-input.js   151 KiB              1  [emitted]  components/bet-input
  components/current-bet.js  84.3 KiB              2  [emitted]  components/current-bet
components/upcoming-game.js  28.8 KiB              3  [emitted]  components/upcoming-game
     components/user-bet.js  30.2 KiB              4  [emitted]  components/user-bet
             pages/admin.js   271 KiB              5  [emitted]  pages/admin
             pages/index.js   187 KiB  6, 1, 2, 3, 4  [emitted]  pages/index
             pages/login.js  99.3 KiB              7  [emitted]  pages/login
          pages/password.js  96.6 KiB              8  [emitted]  pages/password
             pages/table.js   227 KiB              9  [emitted]  pages/table
                  server.js   821 KiB              0  [emitted]  app
       server.manifest.json  1.03 KiB                 [emitted]  
 + 10 hidden assets
Entrypoint app = server.js server.js.map
ℹ Full static generation activated                                                     20:19:39
ℹ Generating output directory: dist/                                                   20:19:39
ℹ Generating pages with full static mode                                               20:19:39
✔ Generated route "/"                                                                  20:19:40
✔ Generated route "/password"                                                          20:19:41
✔ Generated route "/table"                                                             20:19:41
✔ Generated route "/admin"                                                             20:19:41
✔ Generated route "/login"                                                             20:19:41
✔ Client-side fallback created: 200.html                                               20:19:41
✔ Static manifest generated                                                            20:19:41
```

**Reply by r9mp** [reply]
Ok, this is stange.
Can you show the `nuxt.config.ts` ?
Did you put the console.log before or after the resquest execution (it should be before, or else if the request breaks, console log breaks too).

**Reply by imduchy** [reply]
@r9mp I put the `console.log` in front of the request.

```js
// nuxt.config.js

import colors from 'vuetify/es5/util/colors';

export default {
  target: 'static',
  head: {
    // ...
  },
  router: {
    middleware: ['auth', 'groupData'],
  },
  loading: {
    color: '#FBC02D',
    height: '5px',
    throttle: 200,
  },
  css: [],
  plugins: [{ src: '~/plugins/showAlert.ts', mode: 'client' }],
  components: true,
  buildModules: ['@nuxt/typescript-build', '@nuxtjs/vuetify'],
  modules: [
    '@nuxtjs/axios',
    '@nuxtjs/auth-next',
  ],
  auth: {
    // ...
  },
  axios: {
    baseURL: process.env.BASE_URL || 'http://localhost:3003/api',
    credentials: true,
  },
  vuetify: {
    // ...
  },
  build: {
    babel: {
      plugins: [
        ['@babel/plugin-proposal-private-methods', { loose: true }],
        ['@babel/plugin-proposal-private-property-in-object', { loose: true }],
      ],
    },
  },
};
```

**Reply by r9mp** [reply]
@Duchynko 
Hi, sorry I'm answering late!
I created this repo: https://github.com/r9mp/nuxtjs-discussions-static-api-error
The way everything work seems ok to me: "target: static" is set, so the console.log and the data are only fetch once, and this, whatever the fact that you are using fetch or asyncData.
If it is not the case with you project, as it seems to be, there must be a bug or a specific comportment that I do not know. Please try to reproduce it from this repo. If you succeed, either create a bug report or ask me here (but I'm not in the team core so...).
Good luck! And sorry for my late answer.

**Reply by imduchy** [reply]
@r9mp thanks for taking your time and trying to help. I really appreciate it. Your example is, however, different from my use case. Your API call doesn't require any dynamic parameters. As I mentioned in the description,

> ... but what about API calls that require dynamic query parameters?
> ...the API call requires some dynamic values (mostly coming from vuex store).

The API call in the `fetch` hook on my `index.vue` page (see the description) requires dynamic values that are stored in vuex. These values are dispatched to vuex after a user is logged in. Therefore, these pages CAN NOT be statically generated during `nuxt generate`. 

My question, again

> I see that every time I navigate to the index page, this fetch hook is executed and an API call is made. Why is that happening? **Is it because Nuxt can't fetch the data during generation?** Is it because of something else? And if I had an endpoint that doesn't require dynamic query parameters, **how can I make sure that an API call is made on every page load and a user gets always up-to-date data (instead of re-generating the whole site)?**

**Reply by r9mp** [reply]
@Duchynko 

To summarize!

I created a branch `store` on my repo: https://github.com/r9mp/nuxtjs-discussions-static-api-error/blob/store/pages/index.vue

As we are using `target: 'static'`, I'm using `nuxt generate`. If you were not using `target: 'static'`, we would have used `nuxt build`.
In any of these cases, we can use `nuxt start` to start the server.

When I do the `nuxt generate` in my `store` branch, fetch is executed on server and fails. Which, if I understand well what you said, is not your case => this is why I said that maybe there is a bug.
So: "Why is that happening? Is it because Nuxt can't fetch the data during generation?" => It is not happening to me.

"how can I make sure that an API call is made on every page load and a user gets always up-to-date data (instead of re-generating the whole site)?" => For this one, you can set `fetchOnServer` to `false`.  (line 50 in my example, then if you click on "Load Beer" it will work).
=> My example is not the best, because, `$store.beerSlug` is only set though a click. But classical examples are: "is a user connected?", if yes "fetch content is called (store is already set with a good value)", if no "do not call fetch (and redirect guest if needed)". Which could be make like that:
```
async fetch() {
    if (this.$store.state.beerSlug) {
      const result: any = await this.$axios.$get(`https://api.nuxtjs.dev/beers/${this.$store.state.beerSlug}`)

      this.beer = {
        id: result.slug,
        name: result.name,
      }
    }
  },
```

Anyway, for further question, please create a repo, reproduce your problem and explain all steps. If there is a bug, we cannot understand each other I we may have different result for the same code.
